File: 2_Custom_Framerate/c++/main.cpp
#include <opencv2/opencv.hpp>
#include <chrono>

// -- Camera Parameters --
// Camera Index
const int CAM_IDX = 0;
// Resolution
const int WIDTH = 1600;
const int HEIGHT = 1200;

// Framerate
const int FRAMERATE = 60;

// Custom framerate
const int CUSTOM_FRAMERATE = 10;

// Interval in milliseconds to achieve custom framerate
const int FRAME_INTERVAL = 1000 / CUSTOM_FRAMERATE;

// Used to set the pixel format to MJPEG/MJPG mode.
const int MJPG = cv::VideoWriter::fourcc('M', 'J', 'P', 'G');

int main() {
    // -- DEVICE SETUP --
    cv::VideoCapture stellarHD = cv::VideoCapture(CAM_IDX, cv::CAP_V4L2); // V4L2 -> Video 4 Linux

    // Set to MJPEG mode, by default idx 0 is YUYV
    // MJPG needs to be set, before resolution. Pixel format is always selected first
    stellarHD.set(cv::CAP_PROP_FOURCC, MJPG);

    stellarHD.set(cv::CAP_PROP_FRAME_WIDTH, WIDTH);
    stellarHD.set(cv::CAP_PROP_FRAME_HEIGHT, HEIGHT);

    // Set framerate
    stellarHD.set(cv::CAP_PROP_FPS, FRAMERATE);

    // (Optional) Disable auto exposure
    stellarHD.set(cv::CAP_PROP_AUTO_EXPOSURE, 1);
    stellarHD.set(cv::CAP_PROP_EXPOSURE, 90);

    // Create Matrix for frame data
    cv::Mat stellarHDFrame;

    // Timing variables
    auto lastFrameTime = std::chrono::high_resolution_clock::now();

    while(true) {
        auto start = std::chrono::high_resolution_clock::now();
        bool success = stellarHD.read(stellarHDFrame);

        if (success) {
            cv::imshow("stellarHD", stellarHDFrame);
        }

        auto elapsed = std::chrono::high_resolution_clock::now() - start;
        int elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count();

        // Calculate remaining time to wait to maintain the desired frame rate
        int waitTime = FRAME_INTERVAL - elapsed_ms;
        waitTime = std::max(1, waitTime); // Ensuring at least 1ms to prevent freezing

        if (cv::waitKey(waitTime) == 'q') {
            break;
        }
    }

    stellarHD.release();
    cv::destroyAllWindows();

    return 0;
}

File: 2_Custom_Framerate/python/main.py
import cv2
import time

# -- Camera Parameters --
# Camera Index
CAM_IDX = 0
# resolution
WIDTH = 1600
HEIGHT = 1200

# Set a desired custom framerate, by reading frames at a custom interval
# custom framerate must be less than the highest supported framerate of camera
CUSTOM_FRAMERATE = 10

# Hardware supported framerate of the camera. Highest for best results.
FRAMERATE = 60

# used to set the pixel format to MJPEG/MJPG mode.
MJPG = cv2.VideoWriter_fourcc(*'MJPG')

# -- DEVICE SETUP --
stellarHD = cv2.VideoCapture(CAM_IDX)

# set to MJPEG mode, by default idx 0 is YUYV
# MJPG needs to be set, before resolution. Pixel format is always selected first
stellarHD.set(cv2.CAP_PROP_FOURCC, MJPG)

stellarHD.set(cv2.CAP_PROP_FRAME_WIDTH, WIDTH)
stellarHD.set(cv2.CAP_PROP_FRAME_HEIGHT, HEIGHT)

# Set framerate
stellarHD.set(cv2.CAP_PROP_FPS, FRAMERATE)

# (Optional) Disable auto exposure
stellarHD.set(cv2.CAP_PROP_AUTO_EXPOSURE, 1)
stellarHD.set(cv2.CAP_PROP_EXPOSURE, 90)

# Calculate delay
frame_interval = int(1000 / CUSTOM_FRAMERATE)  # in milliseconds

# Error Check
if ((stellarHD == None) or (not stellarHD.isOpened())):
    print('\nError - could not open video device.\n')
    exit(0)

while(True):
    start_time = time.time()
    
    success, frame = stellarHD.read()
    if (success):
        cv2.imshow('stellarHD', frame)
    
    # Calculate remaining time to wait to maintain the desired frame rate
    elapsed_time = (time.time() - start_time) * 1000  # convert to ms
    wait_time = max(1, frame_interval - int(elapsed_time))  # calculate how long to wait

    # Use waitKey to introduce the delay and capture key press
    if cv2.waitKey(wait_time) == ord('q'):
        break

stellarHD.release()
cv2.destroyAllWindows()